internal/libp2pclient/client.go

```diff
@@ -35,93 +33,141 @@ type Client struct {
 	}
 
 	// Start a client
 	return &Client{
-		ctx:         ctx,
-		host:        h,
-		self:        h.ID(),
-		peerID:      peerID,
-		peerSenders: make(map[peer.ID]*peerMessageSender),
-		protocols:   []protocol.ID{protoID},
+		ctxLock: newCtxMutex(),
+		host:    h,
+		peerID:  peerID,
+		protoID: protoID,
 	}, nil
 }
 
+// Self return the peer ID of this client
+func (c *Client) Self() peer.ID {
+	return c.host.ID()
+}
+
+// Close resets and closes the network stream if one exists
+func (c *Client) Close() error {
+	err := c.ctxLock.Lock(context.Background())
+	if err != nil {
+		return err
+	}
+	defer c.ctxLock.Unlock()
+
+	if c.stream != nil {
+		c.closeStream()
+	}
+
+	return nil
+}
+
 // SendRequest sends out a request
 func (c *Client) SendRequest(ctx context.Context, msg proto.Message, decodeRsp DecodeResponseFunc) error {
-	sender, err := c.messageSenderForPeer(ctx, c.peerID)
+	err := c.ctxLock.Lock(ctx)
 	if err != nil {
-		log.Debugw("request failed to open message sender", "error", err, "to", c.peerID)
 		return err
 	}
+	defer c.ctxLock.Unlock()
 
-	return sender.sendRequest(ctx, msg, decodeRsp, c.host, c.protocols)
+	err = c.sendMessage(ctx, msg)
+	if err != nil {
+		return fmt.Errorf("cannot sent request: %w", err)
+	}
+
+	if err = c.ctxReadMsg(ctx, decodeRsp); err != nil {
+		c.closeStream()
+
+		return fmt.Errorf("cannot read response: %w", err)
+	}
+
+	return nil
 }
 
 // SendMessage sends out a message
 func (c *Client) SendMessage(ctx context.Context, msg proto.Message) error {
-	sender, err := c.messageSenderForPeer(ctx, c.peerID)
+	err := c.ctxLock.Lock(ctx)
 	if err != nil {
-		log.Debugw("message failed to open message sender", "error", err, "to", c.peerID)
 		return err
 	}
+	defer c.ctxLock.Unlock()
 
-	if err = sender.sendMessage(ctx, msg, c.host, c.protocols); err != nil {
-		log.Debugw("message failed", "error", err, "to", c.peerID)
+	return c.sendMessage(ctx, msg)
+}
+
+func (c *Client) sendMessage(ctx context.Context, msg proto.Message) error {
+	err := c.prepStreamReader(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err = p2putil.WriteMsg(c.stream, msg); err != nil {
+		c.closeStream()
 		return err
 	}
 
 	return nil
 }
 
-func (c *Client) peerSender(peerID peer.ID) *peerMessageSender {
-	c.sendersLock.Lock()
-	defer c.sendersLock.Unlock()
+func (c *Client) prepStreamReader(ctx context.Context) error {
+	if c.stream == nil {
+		nstr, err := c.host.NewStream(ctx, c.peerID, c.protoID)
+		if err != nil {
+			return err
+		}
 
-	ms, ok := c.peerSenders[peerID]
-	if ok {
-		return ms
+		c.r = msgio.NewVarintReaderSize(nstr, network.MessageSizeMax)
+		c.stream = nstr
 	}
-	ms = &peerMessageSender{
-		peerID:  peerID,
-		ctxLock: newCtxMutex(),
-	}
-	c.peerSenders[peerID] = ms
-	return ms
+
+	return nil
 }
 
-func (c *Client) messageSenderForPeer(ctx context.Context, peerID peer.ID) (*peerMessageSender, error) {
-	ms := c.peerSender(peerID)
+func (c *Client) closeStream() {
+	_ = c.stream.Reset()
+	_ = c.r.Close()
```

Comment:
I think this `c.r.Close()` may be unnecessary since I think `c.stream.Reset()` takes care of that.  Can this be confirmed?

Reply:
I think you are right. We create the `ReadCloser` from the stream in `c.r = msgio.NewVarintReaderSize(nstr, network.MessageSizeMax)`, and `stream.Reset()` closes the stream (and underlying readers) and tells the other end to close its own side.